src/main/haskell/language-kore/src/Data/Kore/Predicate/Predicate.hs

```diff
@@ -0,0 +1,322 @@
+not be used outside of that.
+
+We should consider deleting this and implementing the functionality otherwise.
+-}
+unwrapPredicate :: Predicate level var -> PureMLPattern level var
+unwrapPredicate (GenericPredicate p) = p
+
+-- TODO(virgil): This is wrong, since, e.g., a symbol application
+-- can be a predicate, but should be good enough for now. Fix this.
+isPredicate :: PureMLPattern level var -> Bool
+isPredicate = cata isPredicateOneLevel
+
+isPredicateOneLevel :: Pattern level var Bool -> Bool
+isPredicateOneLevel (AndPattern And {andFirst, andSecond}) =
+    andFirst && andSecond
+isPredicateOneLevel (ApplicationPattern _)               = False
+isPredicateOneLevel (BottomPattern _)                    = True
+isPredicateOneLevel (CeilPattern _)                      = True
+isPredicateOneLevel (CharLiteralPattern _)               = False
+isPredicateOneLevel (DomainValuePattern _)               = False
+isPredicateOneLevel (EqualsPattern _)                    = True
+isPredicateOneLevel (ExistsPattern Exists {existsChild}) = existsChild
+isPredicateOneLevel (FloorPattern _)                     = True
+isPredicateOneLevel (ForallPattern Forall {forallChild}) = forallChild
+isPredicateOneLevel (IffPattern Iff {iffFirst, iffSecond}) =
+    iffFirst && iffSecond
+isPredicateOneLevel (ImpliesPattern Implies {impliesFirst, impliesSecond}) =
+    impliesFirst && impliesSecond
+isPredicateOneLevel (InPattern _)                        = True
+isPredicateOneLevel (NextPattern _)                      = False
+isPredicateOneLevel (NotPattern Not {notChild})          = notChild
+isPredicateOneLevel (OrPattern Or {orFirst, orSecond})   = orFirst && orSecond
+isPredicateOneLevel (RewritesPattern _)                  = False
+isPredicateOneLevel (StringLiteralPattern _)             = False
+isPredicateOneLevel (TopPattern _)                       = True
+isPredicateOneLevel (VariablePattern _)                  = False
+
+{-|'PredicateFalse' is a pattern for matching 'bottom' predicates.
+-}
+pattern PredicateFalse :: Predicate level var
+
+{-|'PredicateTrue' is a pattern for matching 'top' predicates.
+-}
+pattern PredicateTrue :: Predicate level var
+
+pattern PredicateFalse <- GenericPredicate(Bottom_ _)
+pattern PredicateTrue <- GenericPredicate(Top_ _)
+
+{--| 'makeMultipleAndPredicate' combines a list of Predicates with 'and',
+doing some simplification.
+--}
+makeMultipleAndPredicate
+    ::  ( MetaOrObject level
+        , Given (MetadataTools level)
+        , SortedVariable var
+        , Show (var level))
+    => [Predicate level var]
+    -> (Predicate level var, PredicateProof level)
+makeMultipleAndPredicate =
+    foldl'
+        (\(cond1, _) cond2 -> makeAndPredicate cond1 cond2)
+        (makeTruePredicate, PredicateProof)
+
+{--| 'makeAndPredicate' combines two Predicates with an 'and', doing some
+simplification.
+--}
+makeAndPredicate
+    -- TODO(virgil): Group these constraints in a class.
```

Comment:
Why not grouping these in a class, or even better a type (like `ShowMetaOrObject` in `MetaOrObject`), now?